File: src/can/splitting/encrypter.js
/*
Cryptocan - a minimalistic, backward-compatible, and failure resistant CAN-bus encryption
Copyright (C) 2019 Till Zimmermann

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110, USA
*/

//this receives one packet, and sends the ID and Payload encrypted with aes-256-ecb as two frames
//on a specified ID.
//Two sent frames are only identified using their order.
//No Handling for standard/extended-frames exists so far

//our key
var key = new Buffer("3e5595a9e7b764b92985eaff8a448c0e", 'hex')

//Socketcan Wrapper.
var socketcan = require('socketcan');

//OpenSSL Wrapper.
const crypto = require('crypto');

//create Cipher
cipher = crypto.createCipher("aes-256-ecb", key)
cipher.setAutoPadding(false)

//where encrypted messages are being sent.
var encrypted_id = 1

/* Feature:
Frames can be sent in two ways. Either with the same id,
or with two ids for the first resp. second frame.
If two IDs can be used, the transport will be significantly more stable,
as it's clear whether the frame is the first or the second part of
a original frame.
If this is set to false, both are sent on the same id.
Please note that the receiver needs to have the same setting!
*/
var second_id = 2;

//whether the encrypted message are being sent in extended frames.
var encrypted_ext = true;

//Create both interfaces
var plainChannel = socketcan.createRawChannel("can1", true);
var cipherChannel = socketcan.createRawChannel("can0", true);
plainChannel.start();
cipherChannel.start();


var only_id = 12342
var inc_id = false;

plainChannel.addListener("onMessage", function(msg){
    frame_recv = BigInt(msg.ts_sec*1000000000+msg.ts_usec*1000)
    if(only_id!==false && msg.id!=only_id){
        return
    }
    if(inc_id){
        msg.id++;
    }

    //allocate a buffer for two frames
    var fullmessage = Buffer.alloc(16)

    //copy over the id, starting at byte 0
    fullmessage.writeUInt32BE(msg.id);

    //copy over the length, starting at byte15
    fullmessage.writeUInt8(msg.data.length,15);

    //copy over the payload, starting at 4 (could start
    //earlier if we'd had only standard-frames, but the space is free
    //anyways)
    msg.data.copy(fullmessage,4,0,msg.data.length);

    //copy over whether the incoming frame was an extended frame
    fullmessage[14] = msg.ext ? 1 : 0;
    
    //now we have the full message assembled
    //console.log(fullmessage)

    //encrypt it!
    var ciphertext = cipher.update(fullmessage);
    //console.log(ciphertext)

    //and split into two.
    var msg1_data = Buffer.alloc(8);
    var msg2_data = Buffer.alloc(8);
    ciphertext.copy(msg1_data, 0, 0, 8)
    ciphertext.copy(msg2_data, 0, 8, 16)
    var msg1={id:encrypted_id, ext:encrypted_ext, data:msg1_data}
    var msg2={id:(second_id===false ? encrypted_id : second_id), ext:encrypted_ext, data:msg2_data};
    cipherChannel.send(msg1);
    cipherChannel.send(msg2);
    frame_sent = BigInt(msg1.ts_sec*1000000000+msg1.ts_usec*1000)

});


// Ciphertext Layout:
// |ID3|ID2|ID1|ID0|DATA7|DATA6|DATA5|DATA4|DATA3|DATA2|DATA1|DATA0|0|0|EID|DLC

File: src/can/splitting/encrypter.c
/*
Cryptocan - a minimalistic, backward-compatible, and failure resistant CAN-bus encryption
Copyright (C) 2019 Till Zimmermann

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110, USA
*/
#include <errno.h>
#include <inttypes.h>
#include <linux/can.h>
#include <linux/can/raw.h>
#include <net/if.h>
#include <sodium.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <time.h>
#include <unistd.h>
#define AES_BITS 128
#define CRYPTO_CIPH EVP_aes_128_ecb
#include <openssl/conf.h>
#include <openssl/err.h>
#include <openssl/evp.h>

void printFrame(char* note, struct can_frame* f) {
    fprintf(stderr, "%s: ID: %ul, Len: [%d]/[%d], %02x %02x %02x %02x %02x %02x %02x %02x\n", note, f->can_id, f->can_dlc, f->__res0, f->data[0], f->data[1], f->data[2], f->data[3], f->data[4], f->data[5], f->data[6], f->data[7]);
}
void printHex(char* what, uint8_t len) {
    for(int i = 0; i < len; i++) {
        printf("%02X ", what[i]);
    }
    printf("\n");
}

int main(int argc, char** argv) {
    int plainChannel = socket(PF_CAN, SOCK_RAW, CAN_RAW);
    int cipherChannel = socket(PF_CAN, SOCK_RAW, CAN_RAW);
    struct sockaddr_can addr;
    struct ifreq ifr;

    int cipher1_id, cipher2_id;
    cipher1_id = cipher2_id = 268435456 | CAN_EFF_FLAG;
    if(plainChannel < 0) {
        fprintf(stderr, "Couldnt connect: %d\n", plainChannel);
    }
    if(cipherChannel < 0) {
        fprintf(stderr, "Couldnt connect: %d\n", cipherChannel);
    }

    // Channel where plain messages are being received
    strcpy(ifr.ifr_name, (getenv("CAN_RX") ? getenv("CAN_RX") : "can0"));
    ioctl(plainChannel, SIOCGIFINDEX, &ifr);
    addr.can_family = AF_CAN;
    addr.can_ifindex = ifr.ifr_ifindex;
    bind(plainChannel, (struct sockaddr*)&addr, sizeof(addr));

    strcpy(ifr.ifr_name, (getenv("CAN_TX") ? getenv("CAN_TX") : "can1"));
    ioctl(cipherChannel, SIOCGIFINDEX, &ifr);
    addr.can_family = AF_CAN;
    addr.can_ifindex = ifr.ifr_ifindex;
    bind(cipherChannel, (struct sockaddr*)&addr, sizeof(addr));

    // generate random key
    //     uint8_t kRaw[AES_BITS/8]={0};
    //     randombytes_buf(kRaw, AES_BITS/8);
    uint8_t kRaw[] = "NOT A REAL KEY!";

    EVP_CIPHER_CTX* context;
    context = EVP_CIPHER_CTX_new();

    struct can_frame plain;
    struct can_frame cipher_1;
    struct can_frame cipher_2;
    int recv_l;
    while(1) {
        recv_l = recv(plainChannel, &plain, sizeof(plain), 0);
        if(recv_l > 0) {

            if(getenv("VERBOSE")) {
                printFrame("Received", &plain);
            }
            char magic[] = {0xCA, 0xFE, 0xBA, 0xBE, 0xDE, 0xAD, 0xBE, 0xEF};
            if(memcmp(plain.data, magic, 8) == 0) {
                printf("received magic!\n");
                write(cipherChannel, &plain, sizeof(struct can_frame));
                continue;
            }

            uint8_t newFrame[16];
            uint8_t newFrame_encrypted[16];
            int res;
            int reslen = 0;
            int len = 16;
            memset(newFrame, 0, 16);
            uint32_t can_id = plain.can_id & 0b00011111111111111111111111111111;
            newFrame[2] = ((plain.can_id & CAN_EFF_FLAG) == 0) ? 0 : 1;
            memcpy(newFrame + 3, &can_id, 4);
            newFrame[7] = plain.can_dlc;
            memcpy(newFrame + 8, plain.data, (plain.can_dlc > 8 ? 8 : plain.can_dlc));
            EVP_EncryptInit_ex(context, CRYPTO_CIPH(), NULL, kRaw, 0);
            EVP_CIPHER_CTX_set_padding(context, 0);
            EVP_EncryptUpdate(context, newFrame_encrypted, &reslen, newFrame, len);
            if(reslen != len) {
                printf("Failed to encrypt. Encrypted %d bytes, should be %d\n", reslen, len);
                exit(1);
            }

            // should do nothing!
            EVP_EncryptFinal_ex(context, newFrame_encrypted, &reslen);

            // cleanup
            EVP_CIPHER_CTX_cleanup(context);
            cipher_1.can_id = cipher1_id;
            cipher_2.can_id = cipher2_id;
            cipher_1.can_dlc = cipher_2.can_dlc = 8;
            memcpy(cipher_1.data, newFrame_encrypted, 8);
            memcpy(cipher_2.data, newFrame_encrypted + 8, 8);
            if(getenv("VERBOSE")) {
                printFrame("About to Send encrypted 1", &cipher_1);
            }
            if(getenv("VERBOSE")) {
                printFrame("About to Send encrypted 2", &cipher_2);
            }
            res = write(cipherChannel, &cipher_1, sizeof(struct can_frame));
            if(getenv("VERBOSE")) {
                fprintf(stderr, "write ret: %d, %d, %s\n", res, errno, strerror(errno));
            }
            res = write(cipherChannel, &cipher_2, sizeof(struct can_frame));
            if(getenv("VERBOSE")) {
                fprintf(stderr, "write ret: %d, %d, %s\n", res, errno, strerror(errno));
            }
        }
        if(getenv("VERBOSE")) {
            fprintf(stderr, "Handled 1 Frame\n");
        }
    }
}
